### ISSUE TITLE
[plugin request] Telegram

### ISSUE THREAD

**Issue by nylonee** [hangoutsbot/hangoutsbot#232]
This could either be done using the chatbridge framework (See PR #209) or in a standalone plugin.

**Comment by nylonee** [comment]
This. https://github.com/eternnoir/pyTelegramBotAPI

**Comment by FrNico** [comment]
Since last hangout upgrade, I see lots of enthousiam for Telegram, plan is to make a similar plugin to that Slack one? so people can attend same room using either Telegram or Hangout app? That'll allow a smooth migration for some room from Hangout to Telegram...

**Comment by endofline** [comment]
***Disclaimer: No support will be provided for the following Telegram plugin - in its present state, it is only a little bit more than a Proof-of-Concept***

Utilising the [Telegram Bot API](https://core.telegram.org/bots/api) it's possible to register a simple bot that can relay messages in a group by setting `/setprivacy` to `disable`. 

There are some caveats:

* Telegram Bot API has [**no picture downloading capability**](https://twitter.com/telegram/status/614468951926509568)
* It appears that all html formatting will not survive the message relay - so text-only, like Slack
* The interaction of the syncrooms, slack and telegram plugins may be too complicated to resolve, since the Slack-Syncroom interaction itself uses a workaround
* Sink configuration is more involved, requiring the identification of the group ID the Telegram Bot resides in (which - interestingly - is a negative number). On the other hand, only a single sink is required to re-map all Telegram groups to their respective Hangouts:
  ```json
  "telegram": [
    {
      "bot_api_key": "98490337:AAGp5pK3f7hd_1HLdlr3VTX3-o8Ziv-aHZM",
      "conversation_map": [
        {
          "telegram": ["<telegram group id, is a negative number>"],
          "hangouts": ["<conv id>"]
        }
      ]
     }
    ]
    ```
  * The `<telegram group id>` can actually be determined from the url for the group in [Telegram Web](https://web.telegram.org/#/im) - the url query string will resemble the following: `?p=g1111111` - just take the numeric part and add a negative sign i.e. `-1111111'
* Almost completely rewrites the web-bridge framework components, as **[Telegram web-hooks](https://core.telegram.org/bots/api#setwebhook) require the use of HTTPS with a valid certificate** - so we use the [long-polling `getUpdates`](https://core.telegram.org/bots/api#getupdates) method instead afforded by an asyncio-based `Task`

**Comment by FrNico** [comment]
Since Telegram API does not have picture download capability, can we simple copy URL of image starting with https://lh3.googleusercontent.com/blablablabla/xxxxx.jpg
I guess result will be that Telegram will show that image in preview.

**Comment by endofline** [comment]
@frnico that is exactly what happens for picture attachments from hangouts to telegram - the lack of picture downloading is from telegram to hangouts - there is no way of retrieving a picture uploaded via Telegram at the moment

**Comment by pleasantone** [comment]
What do you think about changing the data structure to
"conversation_map": [
  "<group-id>": [ <hangouts> ]
]

We're kinda using dicts wrong here.  (Said as he's writing the slack integration to match)

**Comment by endofline** [comment]
I like this format, I really do :stuck_out_tongue:  It seems like it will last forever. However, due to the diversity of use-cases with these kind of bridge/relays - which includes forwarding-only applications - the earlier proposed `dict` form allows us to add extra keys like "sync-direction: left-to-right, right-to-left, bidirectional (default)" and gives us room to expand for more applications. That's my two-cents

**Comment by pleasantone** [comment]
OK, so you want me to use what's proposed here? I'm fine with it, if that's what you want.

**Comment by endofline** [comment]
The original one noted, though clunky, with a slight modification (changed "telegram" to "external") - here is a contrived example of possible future usage:
```json
    "conversation_map": [
      {
        "external": ["<group id", "group id", "group id", ...],
        "hangouts": ["<conv id>", "<conv id>", "<conv id>", ...],
        "direction": "external-to-internal",
        "autotranslate": true
      },
      {
        <additional conversations>
      }
    ]
```

**Comment by endofline** [comment]
I'm open to suggestions on this, and impose the constraint that the implementation has to preserve the possibility of many-to-many mapping and that it must be expressible in json. In your second example, json won't be able to store that unless you're referring to an internal in-memory dict representation for faster lookups using a frozenset or tuple - the internal representation is at your discretion for specific plugins.

**Comment by pleasantone** [comment]
Excellent point, tuples aren't json, thanks!

**Comment by nylonee** [comment]
Also see alternate project [telesync](https://github.com/makiftasova/hangoutsbot/blob/staging/hangupsbot/plugins/telesync.py) by @makiftasova

**Comment by nylonee** [comment]
Closing this since telesync has been included in the code since 2.7.8
